libsolutil/LP.cpp

```diff
@@ -0,0 +1,805 @@
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolutil/LP.h>
+
+#include <libsolutil/CommonData.h>
+#include <libsolutil/CommonIO.h>
+#include <libsolutil/StringUtils.h>
+#include <libsolutil/LinearExpression.h>
+#include <liblangutil/Exceptions.h>
+
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/transform.hpp>
+#include <range/v3/view/filter.hpp>
+#include <range/v3/view/tail.hpp>
+#include <range/v3/view/iota.hpp>
+#include <range/v3/algorithm/all_of.hpp>
+#include <range/v3/algorithm/any_of.hpp>
+#include <range/v3/algorithm/max.hpp>
+#include <range/v3/algorithm/count_if.hpp>
+#include <range/v3/iterator/operations.hpp>
+
+#include <boost/range/algorithm_ext/erase.hpp>
+
+#include <optional>
+#include <stack>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::util;
+
+using rational = boost::rational<bigint>;
+
+
+namespace
+{
+
+/// Disjunctively combined two vectors of bools.
+inline std::vector<bool>& operator|=(std::vector<bool>& _x, std::vector<bool> const& _y)
+{
+	solAssert(_x.size() == _y.size(), "");
+	for (size_t i = 0; i < _x.size(); ++i)
+		if (_y[i])
+			_x[i] = true;
+	return _x;
+}
+
+/**
+ * Simplex tableau.
+ */
+struct Tableau
+{
+	/// The factors of the objective function (first row of the tableau)
+	LinearExpression objective;
+	/// The tableau matrix (equational form).
+	std::vector<LinearExpression> data;
+};
+
+
+/// Adds slack variables to remove non-equality costraints from a set of constraints
+/// and returns the data part of the tableau / constraints.
+/// The second return variable is true if a the original input had any equality constraints.
+pair<vector<LinearExpression>, bool> toEquationalForm(vector<Constraint> _constraints)
+{
+	size_t varsNeeded = static_cast<size_t>(ranges::count_if(_constraints, [](Constraint const& _c) { return !_c.equality; }));
+	if (varsNeeded > 0)
+	{
+		size_t columns = _constraints.at(0).data.size();
+		size_t varsAdded = 0;
+		for (Constraint& constraint: _constraints)
+		{
+			solAssert(constraint.data.size() == columns, "");
+			constraint.data.resize(columns + varsNeeded);
+			if (!constraint.equality)
+			{
+				constraint.equality = true;
+				constraint.data[columns + varsAdded] = bigint(1);
+				varsAdded++;
+			}
+		}
+		solAssert(varsAdded == varsNeeded);
+	}
+
+	vector<LinearExpression> data;
+	for (Constraint& c: _constraints)
+		data.emplace_back(move(c.data));
+
+	return make_pair(move(data), varsNeeded < _constraints.size());
```

Comment:
Previously this was `varsNeeded > 0`, and I think this is the fix for the issue @bshastry found.